src/main/scala/workflow/AutoCacheRule.scala

```diff
@@ -0,0 +1,531 @@
+package workflow
+
+import breeze.linalg.{max, DenseVector, DenseMatrix}
+import nodes.util.Cacher
+import org.apache.spark.util.SparkUtilWrapper
+import pipelines.Logging
+import workflow.AutoCacheRule.{GreedyCache, AggressiveCache, CachingStrategy}
+
+case class Profile(ns: Long, rddMem: Long, driverMem: Long) {
+  def +(p: Profile) = Profile(this.ns + p.ns, this.rddMem + p.rddMem, this.driverMem + p.driverMem)
+}
+
+case class SampleProfile(scale: Long, profile: Profile)
+
+class AutoCacheRule(cachingMode: CachingStrategy) extends Rule with Logging {
+
+  /**
+   * Get the node weights: estimates for how many passes an instruction will make over its input dependencies
+   */
+  def getNodeWeights(instructions: Seq[Instruction]): Map[Int, Int] = {
+    instructions.zipWithIndex.map {
+      case (TransformerApplyNode(transformer, _), i) => instructions(transformer) match {
+        case node: WeightedNode => (i, node.weight)
+        case _ => (i, 1)
+      }
+      case (EstimatorFitNode(estimator, _), i) => instructions(estimator) match {
+        case node: WeightedNode => (i, node.weight)
+        case _ => (i, 1)
+      }
+      case (_, i) => (i, 1)
+    }.toMap
+  }
+
+  /**
+   * Get a map representing the immediate children for each instruction
+   * Note: This doesn't capture how many times each child depended on the instruction
+   */
+  def getImmediateChildrenByInstruction(instructions: Seq[Instruction]): Map[Int, Seq[Int]] = {
+    instructions.indices.map(i => (i, WorkflowUtils.getImmediateChildren(i, instructions))).toMap
+  }
+
+  /**
+   * Get an estimate for how many times each instruction will be executed, assuming
+   * the given set of instructions have their outputs cached
+   */
+  def getRuns(instructions: Seq[Instruction], cache: Set[Int], nodeWeights: Map[Int, Int]): Map[Int, Int] = {
+    val immediateChildrenByInstruction = getImmediateChildrenByInstruction(instructions)
+
+    instructions.indices.foldRight(Map[Int, Int]()) { case (i, runsByIndex) =>
+      if (immediateChildrenByInstruction(i).isEmpty) {
+        runsByIndex + (i -> 1)
+      }
+      else {
+        val runs = immediateChildrenByInstruction(i).map(j => if (cache.contains(j)) {
+          nodeWeights(j)
+        } else {
+          nodeWeights(j) * runsByIndex(j)
+        }).sum
+        runsByIndex + (i -> runs)
+      }
+    }
+  }
+
+  /**
+   * Get the initial set of what instructions will have their results effectively cached
```

Comment:
So in a sense, the Transformers themselves are ALWAYS cached because they always take up driver memory, it's the output of a transformer being applied to data that isn't necessarily cached. When you see a `TransformerNode` as an instruction, it just signals to store that transformer in the registers, not to apply it to anything. Likewise for `EstimatorNodes` and estimators.